### ISSUE TITLE
Segmentation fault on npc::execute_action

### ISSUE THREAD

**Issue by Gaunah** [CleverRaven/Cataclysm-DDA#38777]
# Describe the bug

I have 2 NPCs in my base and put them on guard to travel north.
If I come back south near the base a segfault is triggered.

# Steps To Reproduce

1. Load the attached save "Creekside".
2. Go south to wards the evac shelter.
3. segfault

# Expected behavior

No crash

# Versions and configuration

 - OS: Windows
    - OS Version: 10.0 1903
- Game Version: 0.D-13130-g4acc2ad [64-bit]
- Graphics Version: Tiles
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food]
]

# Additional context
![SegFault_npc_execute_action](https://user-images.githubusercontent.com/7916253/76690368-94953280-663f-11ea-87e1-e3666038b0c0.png)

[crash.log](https://github.com/CleverRaven/Cataclysm-DDA/files/4333340/crash.log)
[Creekside.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/4333341/Creekside.zip)

**Comment by Gaunah** [comment]
**More Info:**
It only crashes if I'm mounted on the Horse.
If I dismount and walk south everything is fine.

**Comment by kevingranade** [comment]
Looks pretty straightforward, it retrieves "patient" here
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/npcmove.cpp#L901
But never verifies that it's not null.
I'm not sure what the horse has to do with it.

**Comment by kevingranade** [comment]
Digging a little more, the only way this should be reachable is if this code executed:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/npcmove.cpp#L1745
Which means it should either be a reference to the player or a NPC, and there should be no opportunity for that reference to be invalidated in between, so game::shared_from() must have returned nullptr.
I can't see a way for that to happen though, since it simply walks the list of NPCs returned by the game class and then uses one of those references to retrieve a pointer to it instead.

**Comment by kevingranade** [comment]
tl;dr we can definitely keep this from happening, but it's definitely a "should never happen" kind of event.

**Comment by Gaunah** [comment]
I tried to dig into it by me self.
Just as you said up until https://github.com/CleverRaven/Cataclysm-DDA/blob/eb08a43fd489a9b598670c84a5c988878839692d/src/npcmove.cpp#L1751 everything is fine.
But only if I'm mounted `g->shared_from( g->u );` returns a nullptr which isn't checked.
And yes we could just check for that and do nothing but I think we would just bury a deeper issue here.
I'm not familiar with the code and I'm a bit stuck at this point.
